source/geod24/concurrency.d

```diff
@@ -1563,3 +1564,172 @@ private:
     self.receiveOnly!(bool);
     assert(x[0] == 5);
 }
+
+
+/*******************************************************************************
+
+    This is similar to go buffered channels
+
+    Params:
+        T = The type of message to deliver. (int, string, struct, ......)
+
+*******************************************************************************/
+
+public class Channel (T)
+{
+    /// closed
+    private bool closed;
+
+    /// lock for queue and status
+    private Mutex mutex;
+
+    /// queue of data
+    private DList!T queue;
+
+    /// Ctor
+    public this ()
+    {
+        this.closed = false;
+        this.mutex = new Mutex;
+    }
+
+
+    /***************************************************************************
+
+        Send data `msg`.
+
+        Params:
+            msg = message to send
+
+        Return:
+            If successful, return true, otherwise return false
+
+    ***************************************************************************/
+
+    public bool send (T msg)
+    {
+        this.mutex.lock();
+        scope (exit)
+            this.mutex.unlock();
+
+        if (this.closed)
+            return false;
+
+        this.queue.insertBack(msg);
+        return true;
+    }
+
+
+    /***************************************************************************
+
+        Return the received message.
+
+        Return:
+            msg = message to receive
+
+    ***************************************************************************/
+
+    public T receive ()
+    {
+        T msg;
+        while (!tryReceive(&msg))
+        {
+            FiberScheduler.yield();
+        }
+        return msg;
+    }
+
+
+    /***************************************************************************
+
+	    Return the received message.
+
+        Params:
+            msg = point of message to send
+
+        Return:
+            If successful, return true, otherwise return false
+
+    ***************************************************************************/
+
+    public bool tryReceive (T* msg)
+    {
```

Comment:
`assert(msg !is null);`

`tryReceive` will segfault if it's called with an empty pointer.

Reply:
I added it